src/sentry/ingest/transaction_clusterer/rules.py

```diff
@@ -92,7 +95,25 @@ def write(self, project: Project, rules: RuleSet) -> None:
     def merge(self, project: Project) -> None:
         """Read rules from all stores, merge and write them back so they all are up-to-date."""
         merged_rules = self.read(project)
-        self.write(project, merged_rules)
+        trimmed_rules = self._trim_rules(merged_rules)
+        self.write(project, trimmed_rules)
+
+    def _trim_rules(self, rules: RuleSet) -> RuleSet:
+        if self.MERGE_MAX_RULES < 0:
+            return rules
+
+        sorted_rules = sorted(rules.items(), key=lambda p: p[1], reverse=True)
+        if self.MERGE_MAX_RULES < len(rules):
+            with sentry_sdk.configure_scope() as scope:
+                scope.set_tag("discarded", len(rules) - self.MERGE_MAX_RULES)
+                scope.set_context(
+                    "clustering_rules_max",
+                    {"num_existing_rules": len(rules), "max_amount": self.MERGE_MAX_RULES},
```

Comment:
Not a blocker, just a thought: We could even list all the discarded rules in the context here. We have trimming on the event to make sure that the list would not get too long.